src/m_config.c

```diff
@@ -2293,6 +2293,15 @@ float M_GetFloatVariable(const char *name)
     return *variable->location.f;
 }
 
+char *GetPrefPath(const char *org, const char *app) {
+    char *ret = SDL_GetPrefPath(org, app);
+    if (!ret) {
+        ret = strdup("/root/.local/share/chocolate-doom/");
```

Comment:
This is completely wrong. I think the best alternative would be to require a command line parameter in this case and hard error out if not provided.

Reply:
Please consider using `M_TempFile()` instead.

Reply:
Also, we have `M_StringDuplicate()` for `strdup()`.

Reply:
Changing to M_StringDuplicate, being honest I am a little bamboozled how this can be called completely wrong when it fixes the bug and has zero impact on existing execution

Reply:
Those are wrong too, SDL allocates with its internal malloc which must be freed with SDL_free. So this would sometimes allocate with game and sometimes with SDL which is very bad (but might work on Linux).

Reply:
Completely wrong is hardcoding root. If this code fails there's just nothing reasonable to do other than get out.

Reply:
Sounds like every working solution is wrong? Is there an alternate solution?

Reply:
The thing is, about M_TempFile, we don't need a temp file, what's causing the bug is that SDL_GetPrefPath is a null pointer, that needs to be a non-null pointer with a directory.

Reply:
Yes, so in NULL case check command line parameter (which you need to add). If not provided, hard error out.

Reply:
> Completely wrong is hardcoding root. If this code fails there's just nothing reasonable to do other than get out.

The existing code completely fails in this case. The reasonable thing to do is not exit with a segfault (like it currently does, literally a memory violation occurs) and actually try and be contructive and fix things.

Reply:
But we don't want a temporary path but something where the bootloader can put configs and saves.

Reply:
Telling new contributors something is "completely wrong" when it fixes a memory violation is a sure way to kill an open source project

Reply:
This is about running Choco as a daemon during boot splash. I don't think we need actual config and saves in that case?

Reply:
Yup, just tested M_TempFile, works great, thanks for your constructive suggestion @fabiangreffrath ! I removed the extra logging also to be historically accurate as you suggested.

Reply:
> Telling new contributors something is "completely wrong" when it fixes a memory violation is a sure way to kill an open source project

Well, posting a solution with a hard-coded Unix path to root's home for a cross-platform project is also a sure way to get a PR rejected. 😉

Reply:
Hardcoding a path owned by root will also go spectacularly wrong if this code happens to be run as a normal user. There would have to be something seriously wrong with the users environment. Instead of a nice error message or crash they would either get a strange "permission denied" error or just failure to save things. Therefore your proposed fix is completely wrong. I don't know of any way to word it nicer than that.

Reply:
"spectacularly wrong" is quite an exaggeration just like "completely wrong" was, you mean the game wouldn't start.

Reply:
The game might start and do unknown amount of wrong and strange this. That is *much worse* than just crashing outright and absolute hell to debug. The principle of least astonishment exists for a reason.

Reply:
Neither me nor @turol are native speakers, so please excuse our lack of fine-tuning our expressions. This was surely not meant to offend you.

Reply:
@fabiangreffrath you are a brilliant maintainer! You suggest alternative solutions rather than blocking or exaggerating problems. Closing PR, switching to gzdoom.

Reply:
> Closing PR, switching to gzdoom.

Please don't! This PR covers the rare case that `SDL_GetPrefPath()` returns `NULL` which is where the fail actually happens. Strictly, it may only happen for homeless daemons, which is the exact case that you describe in #1448 .

If you could consider all the suggestions outlined during this discussion, I believe this PR has a chance to get accepted.

Reply:
> Yes, so in NULL case check command line parameter (which you need to add). If not provided, hard error out.

Here's a proposed solution. Do this check (even though the check is irrelevant since there are no command line parameters in this case), leave the memory problems in, leave the SEGFAULT occur and "hard error" out, genius.

Reply:
This fix is completely wrong, could cause things to go spectacularly wrong and there would have to be something seriously wrong with the users environment to occur. This new code would also make chocolate-doom absolute hell to debug in future. This is much worse. Setting to "/tmp" would lead to breaking the hard-coded string rule, since every path is hard-coded somewhere.

I'll sleep better knowing I'm not gonna waste more hours on this. There's bigger fish to fry in the open source world.

Reply:
I repeat _much worse_ forgot the italics